src/coreclr/jit/llvm.cpp

```diff
@@ -2563,18 +2637,50 @@ void Llvm::lowerCallToShadowStack(GenTreeCall* callNode)
         bool argOnShadowStack = isThis || (isSigArg && !canStoreArgOnLlvmStack(_compiler, corInfoType, clsHnd));
         if (argOnShadowStack)
         {
-            GenTree* lclShadowStack = _compiler->gtNewLclvNode(_shadowStackLclNum, TYP_I_IMPL);
-
             if (corInfoType == CORINFO_TYPE_VALUECLASS)
             {
                 shadowStackUseOffest = padOffset(corInfoType, clsHnd, shadowStackUseOffest);
             }
 
-            GenTreeIntCon* offset    = _compiler->gtNewIconNode(_shadowStackLocalsSize + shadowStackUseOffest, TYP_I_IMPL);
-            GenTree*       slotAddr  = _compiler->gtNewOperNode(GT_ADD, TYP_I_IMPL, lclShadowStack, offset);
-            GenTree*       storeNode =
-                createShadowStackStoreNode(opAndArg.operand->TypeGet(), slotAddr, opAndArg.operand,
-                                corInfoType == CORINFO_TYPE_VALUECLASS ? _compiler->typGetObjLayout(clsHnd) : nullptr);
+            if (opAndArg.operand->TypeGet() == TYP_STRUCT && opAndArg.operand->OperIs(GT_FIELD_LIST))
+            {
+                for (GenTreeFieldList::Use& use : opAndArg.operand->AsFieldList()->Uses())
+                {
+                    CorInfoType fieldCorInfoType = toCorInfoType(use.GetType());
+
+                    // TODO-LLVM: can this happen?
+                    if (fieldCorInfoType == CORINFO_TYPE_VALUECLASS)
+                    {
+                        failFunctionCompilation();
+                    }
+
+                    GenTree*       lclFieldShadowStack = _compiler->gtNewLclvNode(_shadowStackLclNum, TYP_I_IMPL);
+                    GenTreeIntCon* fieldOffset =
+                        _compiler->gtNewIconNode(_shadowStackLocalsSize + shadowStackUseOffest + use.GetOffset(),
+                                                 TYP_I_IMPL);
+                    GenTree* fieldSlotAddr =
+                        _compiler->gtNewOperNode(GT_ADD, TYP_I_IMPL, lclFieldShadowStack, fieldOffset);
+                    GenTree* fieldStoreNode = createShadowStackStoreNode(use.GetType(), fieldSlotAddr, use.GetNode(), nullptr /* assumes use.GetType() is never CORINFO_TYPE_VALUECLASS* see check above */);
```

Comment:
```suggestion
            if (opAndArg.operand->opAndArg.operand->OperIs(GT_FIELD_LIST))
            {
                for (GenTreeFieldList::Use& use : opAndArg.operand->AsFieldList()->Uses())
                {
                    assert(use.GetType() != TYP_STRUCT);

                    GenTree*       lclShadowStack = _compiler->gtNewLclvNode(_shadowStackLclNum, TYP_I_IMPL);
                    GenTreeIntCon* fieldOffset =
                        _compiler->gtNewIconNode(_shadowStackLocalsSize + shadowStackUseOffest + use.GetOffset(),
                                                 TYP_I_IMPL);
                    GenTree* fieldSlotAddr =
                        _compiler->gtNewOperNode(GT_ADD, TYP_I_IMPL, lclShadowStack, fieldOffset);
                    GenTree* fieldStoreNode = createShadowStackStoreNode(use.GetType(), fieldSlotAddr, use.GetNode(), nullptr);
```
Fields will currently never be of `TYP_STRUCT`; we can assert that.
Nit: `lclFieldShadowStack` -> `lclShadowStack`, since it the same shadow stack local as below.

Otherwise LGTM 👍.

Reply:
Thanks!